Clean up storagePath metrics and expose failed paths in non-debug logs

theoilie opened the pull request:
### Description
- Cleans up the way we pass around types and functions to make storagePath metrics work with cluster mode
- Resets storagePath metrics on each re-run so we can track how many files are migrate in each cycle instead of seeing an infinitely increasing number
- Changes failure logs from debug to warn so we can see them by default
- Bumps redis version for in 2 test places that I noticed were missing. This should maybe be separate, but it's just tests so minimal risk


### Tests
I created a user with a track and manually changed their files to legacy paths in the db before verifying that the metric counts are incremented and specific paths are logged in non-debug mode.
1. `A up; A seed clear; A seed create-user; A seed upload-track`
2. `UPDATE "Files" SET "storagePath" = CONCAT('/file_storage', SUBSTRING("storagePath" FROM 20));`
3. screenshot of logs vs `/prometheus_metrics` (expected / manually created failure because the file doesn't exist on disk):
<img width="1723" alt="Screen Shot 2022-12-13 at 5 46 23 AM" src="https://user-images.githubusercontent.com/4657956/207343228-ea325913-0858-4936-ad71-8862e6201c55.png">

<img width="1316" alt="Screen Shot 2022-12-13 at 10 09 04 AM" src="https://user-images.githubusercontent.com/4657956/207411595-ec1ccf11-89dc-4bec-bd2c-96c975769d00.png">



### Monitoring - How will this change be monitored? Are there sufficient logs / alerts?
- Search "migrate" at `/prometheus_metrics` to make sure it's showing success and failure counts. It turns out stage UM has some legacy files, so check there (I deployed there early for testing)
- Monitor for the errors:
  - "Failed to copy some legacy files"
  - "could not remove mismatched CID file"
  - "Error fixing fileRecord"

SidSethi commented on creator-node/src/services/prometheusMonitoring/prometheus.constants.ts:
```diff
@@ -227,10 +227,12 @@ export const METRIC_LABELS = Object.freeze({
   [METRIC_NAMES.FILES_MIGRATED_FROM_CUSTOM_PATH_GAUGE]: {
-    result: ['success', 'failure']
+    result: ['success'],
+    errMsg: []
```
really dangerous to have open-ended prom metric labels - this is probably fine as the expected cardinality is low, but smth to think about
i understand its difficult to pre-define these since you're not sure exactly how each error will manifest

theoilie replied:
yeah it is a bit of a risk. I've been checking the possible errors and there aren't many so the main concern is if I missed redacting any CIDs or paths then that would make for a very high cardinality. maybe it'd be good to have a backup idea of how to purge the metric in case I missed some edge case that increases cardinality too much. cc @jonaylor89 if you know of anything off the top of your head

SidSethi replied:
yeah we can filter metrics in grafana easily, and delete from prometheus server as well if needed
i guess we'll find out haha
tbh i'm not convinced that all this effort is really worth it vs just monitoring the logs. if we really want to ensure we get this data from SPs that don't publish logs, we can always just ask an SP to grep + share some subset or write each error to a file that is exposed at some new route or something super jank
idk this feels like it creates more risk that we now have to manage

SidSethi commented on creator-node/src/utils/cluster/clusterUtilsForPrimary.ts:
```diff
@@ -1,9 +1,24 @@
 import { isClusterEnabled } from './clusterUtils'
 const cluster: Cluster = require('cluster')
 
+// Lazy-load prometheus utils to avoid DiskManager tests throwing errors when loading it too soon doens't allow everything to be mocked properly
```
seems weird haha - am tempted to just let this one go, do you think there's any other way around this?

theoilie replied:
unfortunately I spent too much time on this for this PR and a past issue 😅  some dependency that DiskManager tests try to mock just don't allow it to import certain things. I think I did make a change to DiskManager at one point that fixed this, but it was a very involved change that was also kinda ugly. if it helps, prom-client lazy loads a module in this same way to avoid a different but similar-ish problem ([code](https://github.com/siimon/prom-client/blob/master/lib/cluster.js#L14-L20))

SidSethi replied:
hahaha its not ideal but if it works locally and in prod then w/e

SidSethi commented on creator-node/src/diskManager.ts:
```diff
@@ -757,31 +744,50 @@ const _migrateFilesWithCustomStoragePaths = async (
+            numFilesMigratedSuccessfully++
+          } else {
+            const errWithoutPaths = replaceAll(
+              replaceAll(
```
i like the cleverness here but same point as above, complexity > utility?

theoilie replied:
similar to the other one, I'm not loving this way of visibility but without it I'm worried we won't be able to debug what remaining files aren't migrated on SPs, especially if they don't export logs or aren't in debug mode. easy to remove if it is an issue tho

theoilie commented on creator-node/src/diskManager.ts:
```diff
@@ -525,16 +521,19 @@ async function _copyLegacyFiles(
         copiedPaths.push({ legacyPath, nonLegacyPath })
       } else {
         erroredPaths[legacyPath] = e.toString()
+        const errWithoutPaths = replaceAll(
```
ah well the problem is that that specific error message isn't chosen by us. I think that one in particular is fs.copyFile or something similar, which comes in full form of `Error: ENOENT: no such file or directory, copyfile 'whatever path1' -> 'whatever path2'`. and if it's some other unexpected error it could also include the path so we have to make sure that path doesn't make into prometheus to spam the cardinality

SidSethi replied:
fair - [my above comment applies here too](https://github.com/AudiusProject/audius-protocol/pull/4468#discussion_r1047513559)

theoilie commented:
@SidSethi I updated to address all the path-specific concerns. PR description is also updated, and this is running on stage UM so you can go to https://usermetadata.staging.audius.co/prometheus_metrics and search for "migrate"

SidSethi reviewed (approved):
amazing, thx for linking to stage example - looks great
